### PR TITLE
[js/webgpu] optimize MatmulNBits

### PR DESCRIPTION
**Pull request by qjia7**
### Description

See 2x speedup for phi3 on the integrated intel gpu with this optimization.

The optimization is mainly to store input A's data into local variable instead of loading them from global memory each time when calculate them with B data.

### CODE REVIEW

**Comment by qjia7** [comment]
@guschmue @fs-eire @satyajandhyala Please take a look, thanks.

**Comment by qjia7** [comment]
@guschmue @fs-eire @satyajandhyala The code is ready for review. But I see three bots are failed `ONNX Runtime React Native CI Pipeline`, `ONNX Runtime React Native CI Pipeline (React Native CI ReactNative_CI)`, `ONNX Runtime Web CI Pipeline (Test_web_MultiBrowsers build_onnxruntime_web_windows)` in this PR and another PR #19388. I think the errors are not related with my changes. Please let me know if you know the possible reasons. Thanks.

**Comment by guschmue** [comment]
sure going the right direction - I see 2x on Xe

**Comment by satyajandhyala** [comment]
Is it possible to add improvement into existing code rather than writing new function? I understand refactoring is more demanding. However less code is easier to maintain.

**Comment by qjia7** [comment]
> Is it possible to add improvement into existing code rather than writing new function? I understand refactoring is more demanding. However less code is easier to maintain.

Yes, my final target is to only keep one blockwise program once the new added one supports all features, like zeroPoint as input, remove extra limitations, like  `nBlocksPerCol < maxComputeWorkgroupSizes[0]`.  To speedup the progress, I use a new program instead of modifying the current code.  In my local, I am also experimenting other optimization ways to see the result.  If everything goes well, I will do the refactoring or submit a new optimization way in the following days. Thanks.

**Comment by qjia7** [comment]
@satyajandhyala The refactor is done. Currently, only one version is provided for MatmulNBits. All limitations are removed.

File: js/web/lib/wasm/jsep/webgpu/ops/matmulnbits.ts [hunk 1]
```diff
@@ -138,187 +120,146 @@ export const createMatMulNBitsProgramInfo = (
...
+            ${output.setByIndices(`${output.type.indices}(batch, row, col + local_id.x)`, 'output_value')};
+          }
         }`;
   };
   return {
-    name: useBlockwiseMatMulNBits ? 'BlockwiseMatMulNBits' : 'MatMulNBits',
+    name: 'MatMulNBits',
     shaderCache: {
-      hint: `${attributes.cacheKey};${dimAOuter};${dataType};${inputs.length}`,
+      hint: `${attributes.blockSize};${attributes.bits};${aComponents};${bComponents};${components};${outputNumber};${workgroupSize}`,
```
Comment (satyajandhyala) [inline js/web/lib/wasm/jsep/webgpu/ops/matmulnbits.ts:257]: workgroupSize is a fixed number. Do we still want to add it here?

**Comment by satyajandhyala** [comment]
Does the pr yields 2x perf improvement on Intel GPU while at least keeping perf same or better on Nvidia?

File: js/web/lib/wasm/jsep/webgpu/ops/matmulnbits.ts [hunk 2]
```diff
@@ -60,41 +59,27 @@ const validateInputs = (inputs: readonly TensorView[], attributes: MatMulNBitsAt
...
   const inputShape = inputs[0].dims;
   const aRank = inputShape.length;
-  const nBlocksPerCol = Math.floor((attributes.k + attributes.blockSize - 1) / attributes.blockSize);
   const dimAOuter = inputShape[aRank - 2];
   const dimInner = attributes.k;
   const dimBOuter = attributes.n;
   const batchDims = inputShape.slice(0, aRank - 2);
   const batchSize = ShapeUtil.size(batchDims);
-  const blobSize = (attributes.blockSize / 8) * attributes.bits;
+  const blobSize = inputs[1].dims[2];
```
Comment (satyajandhyala) [inline js/web/lib/wasm/jsep/webgpu/ops/matmulnbits.ts:70]: The documentation probably is not correct. However, it says "Input B is a 2D constant Matrix" and in the box it says
"Input B is stored as uint8_t with shape: [N][n_blocks_per_col][blob_size] in which ..."
https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#commicrosoftmatmulnbits
Not sure if one should expect input B to be 3 dims.

**Review by qjia7 — COMMENTED** [review]
>Does the pr yields 2x perf improvement on Intel GPU while at least keeping perf same or better on Nvidia?

I think so. This PR is a common optimization. Theoretically, it can bring perf improvement on all gpus. On Nvidia, I see about  80 tokens for phi3 on NV RTX 4090. At least, I didn't see any regression.

@satyajandhyala @guschmue Please help merge if no more issues. Thanks.

**Review by satyajandhyala — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Pull request merged**